### Title: Exercice 6.2
### Author: Jerock Kalala
### Date: October 10th 2022
### Professor: Dr. Brett Werner
### Data Scaling and Working with Dataframes

#### 1. Data Normalization

##### Build a Python function that takes in a vector (array) and normalizes it.

In [58]:
# import module
import numpy as np

# explicit function to normalize array
def normalize(arr):
    norm_arr = []
    for i in arr:
        temp = (i - min(arr))/(np.max(arr) - np.min(arr))
        norm_arr.append(temp)
    return norm_arr


# assign array
array_1d = np.array([1, 2, 4, 8, 10, 15])

print("The original Array = ", array_1d)
print("\nThe array after normalization= ", normalize( array_1d))



The original Array =  [ 1  2  4  8 10 15]

The array after normalization=  [0.0, 0.07142857142857142, 0.21428571428571427, 0.5, 0.6428571428571429, 1.0]


In [59]:
print(" the The L1 norm of the array = ", np.linalg.norm(array_1d, ord=1))
print(" the The L2 norm of the array = ", np.linalg.norm(array_1d, ord=2))

 the The L1 norm of the array =  40.0
 the The L2 norm of the array =  20.248456731316587


In [60]:
#Or we can also use this:
from sklearn import preprocessing

x_array = np.array([1, 2, 4, 8, 10, 15])
normalized_arr = preprocessing.normalize([x_array])
print(normalized_arr)

[[0.04938648 0.09877296 0.19754592 0.39509184 0.4938648  0.7407972 ]]


#### 2. Data Standardization

##### Build a Python function that takes in a vector (array) and standardizes it.

In [72]:
#function to calculate mean
def mean(data):
  n = len(data)
  mean = sum(data) / n
  return mean

#Function to calculate variance. It will help us to calculate the standard deviation
def variance(data):
  n = len(data)
  m = sum(data) / n
  deviations = [(x - m) ** 2 for x in data]
  variance = sum(deviations) / n
  return variance

#function to calculate the standard deviation
def stdev(data):
  import math
  var = variance(data) #call variance function
  std_dev = math.sqrt(var)
  return std_dev

print("Mean of the sample is % s " % (mean(array_1d)))
print("Variance of the sample is % s "% (variance(array_1d)))
print("Standard Deviation of the sample is % s "% (stdev(array_1d)))

Mean of the sample is 6.666666666666667 
Variance of the sample is 23.888888888888886 
Standard Deviation of the sample is 4.887626099538393 


In [68]:
# explicit function to standardize an array
def standard(arr):
    norm_arrA = []
    for i in arr:
        temp = (i - mean(arr))/stdev(arr)
        norm_arrA.append(temp)
    return norm_arrA


print("The standardized array is: ")
standard(array_1d)

The standardized array is: 


[-1.159390377099805,
 -0.954792075258663,
 -0.5455954715763789,
 0.2727977357881894,
 0.6819943394704735,
 1.7049858486761837]

#### 3. Working with a Dataframe

In [79]:
import pandas as pd
df = pd.read_csv("E:\Bellevue\Fall_2022\calif_housing_data.csv")
df

housing_median_age  total_bedrooms  households  median_income  \
0                      41           129.0         126         8.3252   
1                      21          1106.0        1138         8.3014   
2                      52           190.0         177         7.2574   
3                      52           235.0         219         5.6431   
4                      52           280.0         259         3.8462   
...                   ...             ...         ...            ...   
20635                  25           374.0         330         1.5603   
20636                  18           150.0         114         2.5568   
20637                  17           485.0         433         1.7000   
20638                  18           409.0         349         1.8672   
20639                  16           616.0         530         2.3886   

       median_house_value  
0                452600.0  
1                358500.0  
2                352100.0  
3                341300.0  
4                342200.0  
...                   ...  
20635             78100.0  
20636             77100.0  
20637             92300.0  
20638             84700.0  
20639             89400.0  

[20640 rows x 5 columns]

##### (a) How many rows does this data set have?     ==>      20640 rows

##### (b) What is the target vector for your model?   ==>     median house value

##### (c) Create a new feature by taking the total bedrooms divided by the number of households. What does this new feature represent?

In [80]:
df['new_feature'] = df['total_bedrooms'] / df['households']

In [82]:
df.head()

housing_median_age  total_bedrooms  households  median_income  \
0                  41           129.0         126         8.3252   
1                  21          1106.0        1138         8.3014   
2                  52           190.0         177         7.2574   
3                  52           235.0         219         5.6431   
4                  52           280.0         259         3.8462   

   median_house_value  new_feature  
0            452600.0     1.023810  
1            358500.0     0.971880  
2            352100.0     1.073446  
3            341300.0     1.073059  
4            342200.0     1.081081

##### (d) Now, create a new data frame that has three features: the median age, median income, and the new feature created in part (c).

In [91]:
new_df = df[['housing_median_age','median_income','new_feature']]
new_df.head()

housing_median_age  median_income  new_feature
0                  41         8.3252     1.023810
1                  21         8.3014     0.971880
2                  52         7.2574     1.073446
3                  52         5.6431     1.073059
4                  52         3.8462     1.081081

##### (e) Take the data frame created in part (d) and apply data standardization to the features.

In [100]:
print("The standardized median age feature is: ")
`standard(new_df['housing_median_age'])`

The standardized array is: 


[0.9821426581785085,
 -0.6070189133741599,
 1.8561815225324763,
 1.8561815225324763,
 1.8561815225324763,
 1.8561815225324763,
 1.8561815225324763,
 1.8561815225324763,
 1.061600736756142,
 1.8561815225324763,
 1.8561815225324763,
 1.8561815225324763,
 1.8561815225324763,
 1.8561815225324763,
 1.8561815225324763,
 1.6972653653772094,
 1.8561815225324763,
 1.8561815225324763,
 1.6972653653772094,
 1.8561815225324763,
 0.9026845796008751,
 1.061600736756142,
 1.8561815225324763,
 1.8561815225324763,
 1.8561815225324763,
 0.9821426581785085,
 1.617807286799576,
 1.8561815225324763,
 1.6972653653772094,
 1.8561815225324763,
 1.617807286799576,
 1.8561815225324763,
 1.5383492082219425,
 1.617807286799576,
 1.7767234439548427,
 1.617807286799576,
 1.5383492082219425,
 1.8561815225324763,
 1.8561815225324763,
 1.8561815225324763,
 1.8561815225324763,
 1.7767234439548427,
 1.617807286799576,
 1.8561815225324763,
 1.8561815225324763,
 1.8561815225324763,
 1.6972653653772094,
 1.1410588153337755

In [101]:
print("The standardized median income feature is: ")
standard(new_df['median_income'])

The standardized median income feature is: 


[2.344765758301686,
 2.3322379635373007,
 1.7826994032844703,
 0.9329675088245615,
 -0.012880995886464651,
 0.0874466378148665,
 -0.11136630682832735,
 -0.3951366495795735,
 -0.9423591469347871,
 -0.09446957523434674,
 -0.35139464349048327,
 -0.3159167709224365,
 -0.4188236564870231,
 -0.6301117581014737,
 -1.028527214284776,
 -0.9188826911998481,
 -0.5767370358700206,
 -0.921409305269976,
 -0.9893646961977925,
 -0.6671161266702227,
 -1.3227198400753,
 -1.135487209920193,
 -1.1294338637105112,
 -0.8896160782208659,
 -0.6688531738434356,
 -0.7721285239599159,
 -0.7427039976015506,
 -1.0857444954145488,
 -1.172912680833963,
 -1.1491730361333858,
 -1.0228949704201156,
 -1.0049454829635815,
 -1.0915872904517196,
 -1.3136661396573415,
 -0.600266129398087,
 -1.2551855514925048,
 -1.459893928965997,
 -1.2950849986832753,
 -0.20564059431997678,
 -0.6742222287424576,
 -0.9332001709305733,
 -1.3609348778859855,
 -1.4978984156041717,
 0.049284237797308686,
 -0.4517222771918142,
 -0.62842734872138

In [102]:
print("The standardized new_feature income feature is: ")
standard(new_df['new_feature'])

The standardized new_feature income feature is: 


[nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan

In [ ]:
for col in new_df.columns:
    new_df[col] = standard(df[col])
print(new_df)

C:\Users\jeroc\AppData\Local\Temp\ipykernel_20788\48039691.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df[col] = standard(df[col])


In [105]:
from sklearn import preprocessing

#standardize feature
#create a scaler
scaler = preprocessing.StandardScaler()

#transform the feature
x = scaler.fit_transform(new_df['housing_median_age'])
print(x)

ValueError: Expected 2D array, got 1D array instead:
array=[41. 21. 52. ... 17. 18. 16.].
Reshape your data either using array.reshape(-1, 1) if your data has a single feature or array.reshape(1, -1) if it contains a single sample.